File: 251003/Habrus/publisher/src/main/java/by/bsuir/distcomp/repository/InMemoryRepository.java
package by.bsuir.distcomp.repository;

import by.bsuir.distcomp.entity.Identifiable;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

public class InMemoryRepository<T extends Identifiable> implements IRepository<T> {
    private final Map<Long, T> storage = new ConcurrentHashMap<>();
    private static final AtomicLong counter = new AtomicLong();

    @Override
    public Optional<T> findById(Long id) {
        return Optional.ofNullable(storage.get(id));
    }

    @Override
    public List<T> findAll() {
        return storage.values().stream().toList();
    }

    @Override
    public T create(T entity) {
        Long id = counter.incrementAndGet();
        entity.setId(id);
        storage.put(id, entity);
        return entity;
    }

    @Override
    public T update(T entity) {
        storage.put(entity.getId(), entity);
        return entity;
    }

    @Override
    public void deleteById(Long id) {
        storage.remove(id);
    }
}

File: 251003/Pankratiev/rest/src/main/kotlin/bsuir/dc/rest/repository/memory/InMemoryRepository.kt
package bsuir.dc.rest.repository.memory

import bsuir.dc.rest.entity.Identifiable
import bsuir.dc.rest.repository.Repository
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

open class InMemoryRepository<T : Identifiable> : Repository<T> {
    private val storage = ConcurrentHashMap<Long, T>()
    private var idCounter = AtomicLong(1)

    override fun findById(id: Long): T = storage[id]
        ?: throw NoSuchElementException()

    override fun findAll(): List<T> = storage.values.toList()

    override fun save(entity: T): T {
        entity.id = idCounter.getAndIncrement()
        storage[entity.id] = entity
        return entity
    }

    override fun update(entity: T): T {
        validateId(entity.id) {
            storage[entity.id] = entity
        }
        return entity
    }

    override fun deleteById(id: Long) {
        validateId(id) {
            storage.remove(id)
        }
    }

    private fun validateId(id: Long, action: () -> Unit) {
        if (storage.containsKey(id)) {
            action()
        } else {
            throw NoSuchElementException()
        }
    }
}
